File: Quick Sort/quick_sort.py
def quick_sort(arr):
    # Базовый случай: массив из 0 или 1 элемента не требует сортировки
    if len(arr) <= 1:
        return arr

    # Выбираем опорный элемент (pivot)
    pivot = arr[0]
    left = []
    right = []
    equal = []

    # Раскладываем элементы по трем спискам
    for x in arr:
        if x < pivot:
            left.append(x)
        elif x > pivot:
            right.append(x)
        else:
            equal.append(x)

    # Рекурсивно сортируем левую и правую часть
    return quick_sort(left) + equal + quick_sort(right)

# Пример использования
numbers = [10, 7, 8, 9, 1, 5]
sorted_numbers = quick_sort(numbers)
print(sorted_numbers)  # Результат: [1, 5, 7, 8, 9, 10]

# Основная идея: мы делим массив на три части и затем вызываем quick_sort для левой и правой части по отдельности.

File: Quick Sort/quick_sort.js
function quickSort(arr) {
  // Базовый случай
  if (arr.length <= 1) return arr;
  
  // Выбираем опорный элемент (median-of-three стратегия для улучшения производительности)
  const pivot = arr[Math.floor(arr.length / 2)];
  
  const left = [];
  const middle = [];
  const right = [];
  
  // Разделяем массив на три части
  for (let x of arr) {
    if (x < pivot) {
      left.push(x);
    } else if (x === pivot) {
      middle.push(x);
    } else {
      right.push(x);
    }
  }
  
  // Рекурсивно сортируем и объединяем
  return [...quickSort(left), ...middle, ...quickSort(right)];
}

// Альтернативная реализация (сортировка на месте)
function quickSortInPlace(arr, low = 0, high = arr.length - 1) {
  if (low < high) {
    const pi = partition(arr, low, high);
    
    quickSortInPlace(arr, low, pi - 1);
    quickSortInPlace(arr, pi + 1, high);
  }
  
  return arr;
}

function partition(arr, low, high) {
  const pivot = arr[high];
  let i = low - 1;
  
  for (let j = low; j < high; j++) {
    if (arr[j] < pivot) {
      i++;
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }
  
  [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
  return i + 1;
}

// Пример использования
console.log(quickSort([64, 34, 25, 12, 22, 11, 90]));
// Результат: [11, 12, 22, 25, 34, 64, 90]
